File: python/src/tdcv2/unicode/alphabets.py
"""Named character sets, and the ranges they are built from.

``alphabet="cyrillic.ru.letters"`` beats writing thirty-three characters out by hand, and it beats
a range: Ё sits outside the alphabetical block in Unicode but inside it in the alphabet, so a
naive ``а``–``я`` range silently omits a letter Russian actually uses.

Every entry is one CODE POINT, not one UTF-16 unit, so a character outside the basic plane stays
whole rather than being drawn as half a surrogate pair.
"""

from __future__ import annotations


def between(start: int, end: int) -> list[str]:
    """Every code point in an inclusive range."""
    if start > end:
        raise ValueError("invalid alphabet range")
    return [chr(cp) for cp in range(start, end + 1)]


def code_points(value: str) -> list[str]:
    """One entry per code point, so characters outside the basic plane stay whole."""
    return list(value)


_LATIN_LOWER = between(ord("a"), ord("z"))
_LATIN_UPPER = between(ord("A"), ord("Z"))
_DIGITS = between(ord("0"), ord("9"))
# Ё spliced into place rather than appended: it belongs between Е and Ж in the alphabet.
_CYR_LOWER = [*between(ord("а"), ord("е")), "ё", *between(ord("ж"), ord("я"))]
_CYR_UPPER = [*between(ord("А"), ord("Е")), "Ё", *between(ord("Ж"), ord("Я"))]

REGISTRY: dict[str, list[str]] = {
    "latin.lower": _LATIN_LOWER,
    "latin.upper": _LATIN_UPPER,
    "latin.letters": [*_LATIN_UPPER, *_LATIN_LOWER],
    "digits.ascii": _DIGITS,
    "digits.fullwidth": between(ord("０"), ord("９")),
    "cyrillic.ru.lower": _CYR_LOWER,
    "cyrillic.ru.upper": _CYR_UPPER,
    "cyrillic.ru.letters": [*_CYR_UPPER, *_CYR_LOWER],
    "greek.letters": code_points("ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρσςτυφχψω"),
    "hebrew.letters": between(ord("א"), ord("ת")),
    "arabic.letters": between(ord("ء"), ord("ي")),
    "kana.hiragana": between(ord("ぁ"), ord("ゖ")),
    "kana.katakana": between(ord("ァ"), ord("ヺ")),
    "cjk.unified.basic": between(ord("一"), ord("鿿")),
    "roman.upper": code_points("IVXLCDM"),
    "roman.lower": code_points("ivxlcdm"),
}


def chars(name: str) -> list[str] | None:
    """``None`` when the name is unknown; callers report it with the list of known names."""
    return REGISTRY.get(name)


def names() -> list[str]:
    return list(REGISTRY.keys())

File: csharp/Tdcv2/Unicode/Alphabets.cs
namespace Tdcv2.Unicode;

/// <summary>
/// The named alphabets a config can ask for by name.
/// </summary>
/// <remarks>
/// Spelled out as explicit code-point ranges rather than looked up through Unicode character
/// properties. Property tables ship with the runtime and change between versions of it, so a config
/// that drew Cyrillic letters could quietly draw a different set of them after an upgrade, and two
/// languages' runtimes would never agree. A range written here means the same thing everywhere,
/// forever.
/// </remarks>
public static class Alphabets
{
    private static readonly Dictionary<string, IReadOnlyList<string>> Registry = Build();

    private static Dictionary<string, IReadOnlyList<string>> Build()
    {
        IReadOnlyList<string> latinLower = Between('a', 'z');
        IReadOnlyList<string> latinUpper = Between('A', 'Z');
        IReadOnlyList<string> digits = Between('0', '9');

        // Ё sits outside the alphabetical block in Unicode but inside it in the alphabet, so it is
        // spliced back into place rather than appended.
        var cyrLower = new List<string>(Between('а', 'е')) { "ё" };
        cyrLower.AddRange(Between('ж', 'я'));
        var cyrUpper = new List<string>(Between('А', 'Е')) { "Ё" };
        cyrUpper.AddRange(Between('Ж', 'Я'));

        // Insertion order is preserved, as it is in the reference: an alphabet is a list to draw
        // from, and the order of that list is what the draw index means.
        return new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal)
        {
            ["latin.lower"] = latinLower,
            ["latin.upper"] = latinUpper,
            ["latin.letters"] = Concat(latinUpper, latinLower),
            ["digits.ascii"] = digits,
            ["digits.fullwidth"] = Between('０', '９'),
            ["cyrillic.ru.lower"] = cyrLower,
            ["cyrillic.ru.upper"] = cyrUpper,
            ["cyrillic.ru.letters"] = Concat(cyrUpper, cyrLower),
            ["greek.letters"] =
                CodePoints("ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩαβγδεζηθικλμνξοπρσςτυφχψω"),
            ["hebrew.letters"] = Between('א', 'ת'),
            ["arabic.letters"] = Between('ء', 'ي'),
            ["kana.hiragana"] = Between('ぁ', 'ゖ'),
            ["kana.katakana"] = Between('ァ', 'ヺ'),
            ["cjk.unified.basic"] = Between('一', '鿿'),
            ["roman.upper"] = CodePoints("IVXLCDM"),
            ["roman.lower"] = CodePoints("ivxlcdm"),
        };
    }

    /// <summary><c>null</c> when the name is unknown; callers report it with the list of known names.</summary>
    public static IReadOnlyList<string>? Chars(string name) =>
        Registry.TryGetValue(name, out IReadOnlyList<string>? chars) ? chars : null;

    public static IReadOnlyList<string> Names() => Registry.Keys.ToList();

    public static IReadOnlyList<string> Between(int start, int end)
    {
        if (start > end)
        {
            throw new ArgumentException("invalid alphabet range");
        }

        var result = new List<string>(end - start + 1);
        for (int cp = start; cp <= end; cp++)
        {
            result.Add(char.ConvertFromUtf32(cp));
        }

        return result;
    }

    /// <summary>One entry per code point, so characters outside the basic plane stay whole.</summary>
    public static IReadOnlyList<string> CodePoints(string value)
    {
        var result = new List<string>();
        for (int i = 0; i < value.Length;)
        {
            int cp = char.ConvertToUtf32(value, i);
            result.Add(char.ConvertFromUtf32(cp));
            i += char.IsHighSurrogate(value[i]) ? 2 : 1;
        }

        return result;
    }

    private static IReadOnlyList<string> Concat(IReadOnlyList<string> a, IReadOnlyList<string> b)
    {
        var result = new List<string>(a.Count + b.Count);
        result.AddRange(a);
        result.AddRange(b);
        return result;
    }
}
